BUILD.md

```diff
@@ -25,178 +10,323 @@ For security related issues please see [SECURITY.md](SECURITY.md).
-
-`make -C py/bitbox02` to generate the protobuf files.
-
-To kick off some api calls:
+Run it with:
 
 ```sh
-./py/send_message.py
+./build-build/bin/simulator
 ```
 
-### Unit tests
+This launches a server simulating the firmware. The send_message tool can connect to it with:
+
+    ./py/send_message.py --simulator
+
+If you choose to create a wallet by restoring a mnemonic, the simulator will automatically use this
+mnemonic:
+
+    boring mistake dish oyster truth pigeon viable emerge sort crash wire portion cannon couple enact box walk height pull today solid off enable tide
+
+
+#### Debugging using the J-Link probe and GDB
+
+The *debug firmware* enables pretty printing of panics over [RTT](https://www.segger.com/products/debug-probes/j-link/technology/about-real-time-transfer/).
 
-We are using CMocka [https://cmocka.org/](https://cmocka.org/) for unit tests. To run the tests, the CMocka library
-needs to be installed on your system.
+Run the following commands to build the debug firmware.
 
-If you're on a Mac, you can use the brew command to install it:
+```sh
+make dockerdev
+make firmware-debug
+```
+
+Run the following command to run the J-Link GDB Server.
 
 ```sh
-brew install cmocka
+make jlink-gdb-server
 ```
 
-Alternatively, you can get CMocka by cloning the git repository and following these instructions:
+> [!IMPORTANT]
+> The J-Link GDB Server must be left running in the background.
+
+Run the following command to connect with telnet to the J-Link GDB Server to
+see the RTT output.
 
 ```sh
-git clone git://git.cryptomilk.org/projects/cmocka.git
-cd cmocka
-mkdir build && cd build
-cmake ..
-make && sudo make install
+make rtt-client
 ```
 
-By default, the library will be installed into /usr/local/lib64 directory under Linux x86\_64.
-If the library is not on the library path by default, you might need to export the following environment variable:
+Run the following command to run GDB. GDB will connect to the J-Link GDB
+server, flash the debug firmware and then start execution from the bootloader
+(as if the device was just plugged in).
 
 ```sh
-export LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:/usr/local/lib64/
+make run-debug
 ```
 
+> [!TIP]
+> After rebuilding the firmware, exit GDB and rerun `run-debug` to flash and reset the device.
+
+> [!TIP]
+> The initial set of GDB commands that are run are specified in the [gdb init
+> script](./scripts/jlink.gdb). You may want to modify it if you are debugging
+> something specific.
+
+> [!TIP]
+> In debug builds you can use the following functions to log:
+> ```c
+> util_log(fmt, args...)
+> ```
+> ```rust
+> use ::util::log::log!(fmt, args...)
+> ```
+> in C you can also format with hex using `util_dbg_hex`:
+> ```c
+> uint8_t arr[] = {1,2};
+> util_log("%s", util_dbg_hex(arr, sizeof(arr)));
+> ```
+> in rust you can format with hex using the built in hex formatter:
+> ```rust
+> let arr = [1, 2];
+> log!("{:02x?}", arr)
```

Comment:
This prints `[01, 02]` which is not nice, especially for long hex strings (e.g. 32 bytes). `log!("{}", hex::encode(arr))` is better.

Reply:
Added that as well. Sometimes I think it is easier if you can count the bytes and then the whitespace is helpful.